### PR TITLE
restore: re-enable online restore of tenants

### PR DESCRIPTION
**Pull request by dt**
Release note: none.
Epic: none.

### CODE REVIEW

**Review by msbutler — APPROVED** [review]
Thanks for doing this! Apologies for being sluggish to get this done.

File: pkg/ccl/backupccl/restore_online.go [hunk 1]
```diff
@@ -171,6 +172,22 @@ func sendAddRemoteSSTWorker(
...
+			// keys, including within the subspace of a tenant span.
+			if entry.ElidedPrefix == execinfrapb.ElidePrefix_Tenant {
+				_, id, err := keys.DecodeTenantPrefix(entry.Span.Key)
+				if err != nil {
+					return err
+				}
+				if tSpan := keys.MakeTenantSpan(id); entry.Span.EndKey.Equal(tSpan.EndKey) {
+					entry.Span.EndKey = append(tSpan.Key, keys.MaxKey...)
+				}
+			}
```
Comment (msbutler) [inline pkg/ccl/backupccl/restore_online.go:190]: er, you have to do this transformation for each backup file entry too. I'll ad this logic to `rewriteSpan`.

**Review by msbutler — CHANGES_REQUESTED** [review]
After a second look, i realized this isn't quite right. happy to take over if you'd like.

File: pkg/ccl/backupccl/restore_online.go (hunk 1 above)
Comment (dt) [reply pkg/ccl/backupccl/restore_online.go:190]: do we not assert those the way we do the entry spans?

Comment (dt) [reply pkg/ccl/backupccl/restore_online.go:190]: put another way: why is this green?

Comment (dt) [reply pkg/ccl/backupccl/restore_online.go:190]: we answered why this is green, which is that we restore the tenant to the same ID.

**Comment by dt** [comment]
@msbutler RFAL

**Comment by dt** [comment]
actually _now_ it is RFAL; fixes the sink tests

File: pkg/ccl/backupccl/restore_online.go [hunk 2]
```diff
@@ -171,11 +172,40 @@ func sendAddRemoteSSTWorker(
...
 			for _, file := range entry.Files {
+				if entry.ElidedPrefix == execinfrapb.ElidePrefix_Tenant {
+					_, id, err := keys.DecodeTenantPrefix(file.BackupFileEntrySpan.Key)
+					if err != nil {
+						return err
+					}
+					if tSpan := keys.MakeTenantSpan(id); file.BackupFileEntrySpan.EndKey.Equal(tSpan.EndKey) {
+						file.BackupFileEntrySpan.EndKey = append(tSpan.Key, keys.MaxKey...)
+					}
+				}
```
Comment (stevendanna) [inline pkg/ccl/backupccl/restore_online.go:204]: [nit] Up to you: Could be worth extracting to a function

File: pkg/ccl/backupccl/restore_online.go [hunk 3]
```diff
@@ -171,11 +172,40 @@ func sendAddRemoteSSTWorker(
...
+				if err := assertCommonPrefix(file.BackupFileEntrySpan, entry.ElidedPrefix); err != nil {
```
Comment (msbutler) [inline pkg/ccl/backupccl/restore_online.go:205]: nit: we could remove the `assertCommonPrefix()` call in `rewriteSpan()`, as we now check the file and the span entry before intersection.

**Review by msbutler — APPROVED** [review]
LGTM, modulo Steven's question about DecodeMVCCValue X ImportEpochs

File: pkg/ccl/backupccl/file_sst_sink.go [hunk 4]
```diff
@@ -316,16 +316,26 @@ func (s *fileSSTSink) copyPointKeys(dataSST []byte) error {
...
 		}
+		v, err := storage.DecodeValueFromMVCCValue(raw)
+		if err != nil {
+			return errors.Wrapf(err, "decoding mvcc value %s", k)
+		}
+
+		// Checksums include the key, but *exported* keys no longer live at that
+		// key once they are exported, and could be restored as some other key, so
+		// zero out the checksum.
+		v.ClearChecksum()
```
[the remark this answers is not in the document]
Comment (dt) [reply pkg/ccl/backupccl/file_sst_sink.go]: MB added a patch to do this.

**Comment by dt** [comment]
Thanks for the review and extra fix @msbutler!

bors r+

**Comment by msbutler** [comment]
merge conflict. fixing.
bors r-

**Pull request merged**